modules/cudaarithm/src/cuda/in_range.cu

```diff
@@ -0,0 +1,153 @@
+}
+
+}  // namespace
+
+void cv::cuda::inRange(InputArray _src,
+                       InputArray _lowerb,
+                       InputArray _upperb,
+                       OutputArray _dst,
+                       Stream& stream) {
+    const GpuMat src = getInputMat(_src, stream);
+
+    const _InputArray::KindFlag skind = _src.kind();
+    const _InputArray::KindFlag lkind = _lowerb.kind();
+    const _InputArray::KindFlag ukind = _upperb.kind();
+
+    const Mat lb = _lowerb.getMat();
+    const Mat ub = _upperb.getMat();
+
+    // lb and ub must both be cv::Scalar
+    CV_Assert(lb.size() == Size(1, 4));
+    CV_Assert(ub.size() == Size(1, 4));
+
+    CV_Assert(lb.depth() == CV_64F);
+    CV_Assert(ub.depth() == CV_64F);
+
+    CV_Assert(lb.channels() == 1);
+    CV_Assert(ub.channels() == 1);
+
+    const int depth = src.depth();
+    const int channels = src.channels();
+
+    GpuMat dst = getOutputMat(_dst, src.size(), CV_8UC1, stream);
+
+    typedef void (*func_t)(const GpuMat& src,
+                           const Mat& lowerb,
+                           const Mat& upperb,
+                           GpuMat& dst,
+                           Stream& stream);
+    static const std::array<std::array<func_t, 7>, 4> funcs = {
+            std::array<func_t, 7>{inRangeImpl<uchar, 1>,
+                                  inRangeImpl<schar, 1>,
+                                  inRangeImpl<ushort, 1>,
+                                  inRangeImpl<short, 1>,
+                                  inRangeImpl<int, 1>,
+                                  inRangeImpl<float, 1>,
+                                  inRangeImpl<double, 1>},
+            std::array<func_t, 7>{inRangeImpl<uchar, 2>,
+                                  inRangeImpl<schar, 2>,
+                                  inRangeImpl<ushort, 2>,
+                                  inRangeImpl<short, 2>,
+                                  inRangeImpl<int, 2>,
+                                  inRangeImpl<float, 2>,
+                                  inRangeImpl<double, 2>},
+            std::array<func_t, 7>{inRangeImpl<uchar, 3>,
+                                  inRangeImpl<schar, 3>,
+                                  inRangeImpl<ushort, 3>,
+                                  inRangeImpl<short, 3>,
+                                  inRangeImpl<int, 3>,
+                                  inRangeImpl<float, 3>,
+                                  inRangeImpl<double, 3>},
+            std::array<func_t, 7>{inRangeImpl<uchar, 4>,
+                                  inRangeImpl<schar, 4>,
+                                  inRangeImpl<ushort, 4>,
+                                  inRangeImpl<short, 4>,
+                                  inRangeImpl<int, 4>,
+                                  inRangeImpl<float, 4>,
+                                  inRangeImpl<double, 4>}};
+
+    const func_t func = funcs.at(channels - 1).at(depth);
+    func(src, lb, ub, dst, stream);
```

Comment:
Please use `CV_Check*()` macro to validate `channels` / `depth` values.

Reply:
Fixed

Reply:
> src.depth()

where is validated?

Reply:
Previously wasn't; I was trying to cover all depths.  I've now realized that it doesn't work for float16, because CUDA doesn't have a float16 vector type.  So I added a CV_Check that the depth is CV_64F or lower